lambdas/replay-cimit-vcs/src/main/java/uk/gov/di/ipv/core/replaycimitvcs/ReplayCimitVcsHandler.java

```diff
@@ -0,0 +1,83 @@
+import software.amazon.lambda.powertools.tracing.Tracing;
+import uk.gov.di.ipv.core.library.annotations.ExcludeFromGeneratedCoverageReport;
+import uk.gov.di.ipv.core.library.cimit.exception.CiPostMitigationsException;
+import uk.gov.di.ipv.core.library.domain.ReplayItem;
+import uk.gov.di.ipv.core.library.domain.ReplayRequest;
+import uk.gov.di.ipv.core.library.exceptions.FailedVcReplayException;
+import uk.gov.di.ipv.core.library.helpers.LogHelper;
+import uk.gov.di.ipv.core.library.persistence.item.VcStoreItem;
+import uk.gov.di.ipv.core.library.service.CiMitService;
+import uk.gov.di.ipv.core.library.service.ConfigService;
+import uk.gov.di.ipv.core.library.verifiablecredential.service.VerifiableCredentialService;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ReplayCimitVcsHandler implements RequestStreamHandler {
+    private static final Logger LOGGER = LogManager.getLogger();
+    private final ConfigService configService;
+    private final CiMitService ciMitService;
+    private final VerifiableCredentialService verifiableCredentialService;
+
+    @SuppressWarnings("unused") // Used by AWS
+    public ReplayCimitVcsHandler(
+            ConfigService configService,
+            CiMitService ciMitService,
+            VerifiableCredentialService verifiableCredentialService) {
+        this.configService = configService;
+        this.ciMitService = ciMitService;
+        this.verifiableCredentialService = verifiableCredentialService;
+    }
+
+    @SuppressWarnings("unused") // Used through dependency injection
+    @ExcludeFromGeneratedCoverageReport
+    public ReplayCimitVcsHandler() {
+        this.configService = new ConfigService();
+        this.ciMitService = new CiMitService(configService);
+        this.verifiableCredentialService = new VerifiableCredentialService(configService);
+    }
+
+    @Override
+    @Tracing
+    @Logging(clearState = true)
+    public void handleRequest(InputStream inputStream, OutputStream outputStream, Context context) {
+        LogHelper.attachComponentIdToLogs(configService);
+        ObjectMapper mapper = new ObjectMapper();
+        ReplayRequest request;
+        try {
+            request = mapper.readValue(inputStream, ReplayRequest.class);
+        } catch (IOException e) {
+            LOGGER.error("Failed to map request to valid replay event", e);
+            throw new FailedVcReplayException(
+                    String.format("Failed to replay VCs to CIMIT because: '%s'", e));
+        }
+        LOGGER.info("Retrieving {} VCs", request.getItems().size());
+        List<VcStoreItem> vcStoreItems = new ArrayList<>();
+        for (ReplayItem item : request.getItems()) {
+            VcStoreItem vcStoreItem =
+                    this.verifiableCredentialService.getVcStoreItem(
+                            item.getUserId().get("S"), item.getCredentialIssuer().get("S"));
```

Comment:
Is this reasonably fast? It's going to make a separate call to DynamoDB for each VC, which could add up quickly...

If it's a problem, you might want to batch them up (i.e. https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchGetItem.html) - these could be the batches that you send to CIMIT as well.

Reply:
I considered this, the current implementation seems quick, although I've only tried in a relatively small table on dev - something to look at when updating to send in batches